src/boost_histogram/_internal/axis.py

```diff
@@ -37,6 +37,31 @@ def __copy__(self):
         other._ax = copy.copy(self._ax)
         return other
 
+    def __getattr__(self, item):
+        if item == "_ax":
+            return Axis.__dict__[item].__get__(self)
+        elif self._ax.metadata is not None and item in self._ax.metadata:
+            return self._ax.metadata[item]
+        elif item == "metadata":
+            return None
+        else:
+            raise AttributeError(
+                "'{}' object has no attribute '{}'".format(type(self).__name__, item)
+            )
+
+    def __setattr__(self, item, value):
+        if item == "_ax":
+            Axis.__dict__[item].__set__(self, value)
+            return
+        elif self._ax.metadata is None:
+            self._ax.metadata = {}
+
+        self._ax.metadata[item] = value
+
+    def __dir__(self):
+        metadata = list(self._ax.metadata) if self.ax.metadata is not None else []
```

Comment:
That only works if `self._ax.metadata` is a dict, right? What if metadata is a number or a string?

Reply:
I see below that you remove the possibility to set `metadata` by hand, so perhaps this is ok then, but it should be made sure that `metadata` is always a dict.

Reply:
Yes, `self._ax.metadata` is basically the equivalent of `self.__dict__` without `self._ax`. `self.metadata` is just an item in that dict, with the (backwards compatible) feature that it defaults to None if not present.